File: leetcode/solutions/6004.FirstUniqueNumber/FirstUnique.go
package solution

// DoubleListNode is a structure of doubly linked list
type DoubleListNode struct {
	val  int
	prev *DoubleListNode
	next *DoubleListNode
}

// FirstUnique represents an object manupilated.
type FirstUnique struct {
	head    *DoubleListNode
	tail    *DoubleListNode
	listMap map[int]*DoubleListNode
}

// Constructor returns an instance of FirstUnique.
func Constructor(nums []int) FirstUnique {
	head := &DoubleListNode{val: -2}
	tail := &DoubleListNode{val: -1, prev: head}
	head.next = tail
	fu := FirstUnique{
		head:    head,
		tail:    tail,
		listMap: make(map[int]*DoubleListNode),
	}

	for _, num := range nums {
		fu.Add(num)
	}

	return fu
}

// ShowFirstUnique returns the first unique number passed by Add() method.
func (fu *FirstUnique) ShowFirstUnique() int {
	return fu.head.next.val
}

// Add simply adds geven number at the end of DoubleListNode if it is unique and set it into listMap to keep it distinct.
// Remove the given number from DoubleListNode if it already exists in listMap.
func (fu *FirstUnique) Add(value int) {
	if n, ok := fu.listMap[value]; ok {
		if n.prev == nil {
			return
		}

		n.prev.next = n.next
		n.next.prev = n.prev
		n.prev, n.next = nil, nil
		return
	}

	node := &DoubleListNode{val: value, prev: fu.tail.prev, next: fu.tail}
	fu.tail.prev.next = node
	fu.tail.prev = node
	fu.listMap[value] = node
}

File: leetcode/solutions/6004.FirstUniqueNumber/FirstUnique.ts
export default class FirstUnique {
  private listMap: Map<number, DoubleListNode | null>;
  private head: DoubleListNode;
  private tail: DoubleListNode;

  constructor(nums: number[]) {
    this.head = { val: -2, next: null, prev: null };
    this.tail = { val: -1, next: null, prev: this.head };
    this.head.next = this.tail;
    this.listMap = new Map<number, DoubleListNode | null>();

    for (const n of nums) {
      this.add(n);
    }
  }

  showFirstUnique(): number {
    return this.head.next!.val;
  }

  add(value: number) {
    if (this.listMap.has(value)) {
      const node = this.listMap.get(value);

      if (!node || !node.prev || !node.next) {
        return;
      }

      node.prev.next = node.next;
      node.next.prev = node.prev;
      this.listMap.set(value, null);

      return;
    }

    const node: DoubleListNode = {
      val: value,
      prev: this.tail.prev,
      next: this.tail,
    };
    this.listMap.set(value, node);
    this.tail.prev!.next = node;
    this.tail.prev = node;
  }
}

type DoubleListNode = {
  val: number;
  prev: DoubleListNode | null;
  next: DoubleListNode | null;
};
